src/util/RtreeBuild.cpp

```diff
@@ -0,0 +1,1133 @@
+      Rtree::BoundingBox b1 = Rtree::createBoundingBox(minXB1, minYB1, maxXB1, maxYB1);
+
+
+      double cost = costFunctionTGS(b0, b1, dim);
+
+      if (splitResult.bestCost == -1 || cost < splitResult.bestCost) {
+        splitResult.bestCost = cost;
+        splitResult.bestDim = dim;
+        splitResult.bestLastElement = currentLastElement;
+        splitResult.bestElement = currentElement;
+        splitResult.bestMinElement = minElement;
+        splitResult.bestMaxElement = maxElement;
+        splitResult.bestIndex = i;
+      }
+    }
+    currentlyAtSTimesI = false;
+  }
+
+  return splitResult;
+}
+
+std::pair<OrderedBoxes, OrderedBoxes> OrderedBoxes::SplitAtBest(const std::string& filePath, size_t S, size_t M, uint64_t maxBuildingRamUsage) {
+  if (this->workInRam) {
+    return this->SplitAtBestInRam(S, M);
+  } else {
+    return this->SplitAtBestOnDisk(filePath, S, M, maxBuildingRamUsage);
+  }
+}
+
+std::pair<OrderedBoxes, OrderedBoxes> OrderedBoxes::SplitAtBestInRam(size_t S, size_t M) {
+  /**
+     * Split the ordered boxes in ram. First determine the best split and then perform it
+   */
+
+  struct SplitResult splitResult = this->GetBestSplit();
+
+  OrderedBoxes split0;
+  OrderedBoxes split1;
+
+  struct SplitBuffersRam splitBuffers;
+
+  splitBuffers.s0Dim0 = std::make_shared<multiBoxWithOrderIndex>();
+  splitBuffers.s0Dim1 = std::make_shared<multiBoxWithOrderIndex>();
+  splitBuffers.s1Dim0 = std::make_shared<multiBoxWithOrderIndex>();
+  splitBuffers.s1Dim1 = std::make_shared<multiBoxWithOrderIndex>();
+
+  splitBuffers.s0SmallDim0 = std::make_shared<multiBoxWithOrderIndex>();
+  splitBuffers.s0SmallDim1 = std::make_shared<multiBoxWithOrderIndex>();
+  splitBuffers.s1SmallDim0 = std::make_shared<multiBoxWithOrderIndex>();
+  splitBuffers.s1SmallDim1 = std::make_shared<multiBoxWithOrderIndex>();
+
+  std::pair<Rtree::BoundingBox, Rtree::BoundingBox> boundingBoxes = PerformSplit(splitResult, splitBuffers, M, S);
+
+  RectanglesForOrderedBoxes rectsD0Split0 = {splitBuffers.s0Dim0, "", splitBuffers.s0SmallDim0};
+  RectanglesForOrderedBoxes rectsD1Split0 = {splitBuffers.s0Dim1, "", splitBuffers.s0SmallDim1};
+  RectanglesForOrderedBoxes rectsD0Split1 = {splitBuffers.s1Dim0, "", splitBuffers.s1SmallDim0};
+  RectanglesForOrderedBoxes rectsD1Split1 = {splitBuffers.s1Dim1, "", splitBuffers.s1SmallDim1};
+  split0.CreateOrderedBoxesInRam(rectsD0Split0, rectsD1Split0, boundingBoxes.first);
+  split1.CreateOrderedBoxesInRam(rectsD0Split1, rectsD1Split1, boundingBoxes.second);
+
+  (*this->rectsD0.rectanglesInRam).clear();
+  (*this->rectsD1.rectanglesInRam).clear();
+  (*this->rectsD0.rectanglesSmall).clear();
+  (*this->rectsD1.rectanglesSmall).clear();
+  (*this->rectsD0.rectanglesInRam).shrink_to_fit();
+  (*this->rectsD1.rectanglesInRam).shrink_to_fit();
+  (*this->rectsD0.rectanglesSmall).shrink_to_fit();
+  (*this->rectsD1.rectanglesSmall).shrink_to_fit();
```

Comment:
This looks rather fishy:
This function never touches those vectors, but afterwards clears them.
If necessary, this clear should be at the end of the function that actually uses those buffers).